### ISSUE TITLE
Unable to resolve service for type 'System.Func`1[Sentry.IHub] when adding app.UseSentryTracing() and running as AWS Lambda

### ISSUE THREAD

**Issue by weare8** [getsentry/sentry-dotnet#1765]
Please mark the type framework used:

- [ X ] ASP.NET Core

Please mark the type of the runtime used:

- [ X ] .NET Core

* Version: 6.0

Please mark the NuGet packages used:

- [ X ] Sentry.AspNetCore

* Version: 3.19.0

Create am AWS Lambda ASP.NET Core Web API Project (https://github.com/aws/aws-lambda-dotnet - Lambda ASP.NET Core Web API - serverless.AspNetCoreWebAPI).
This will provide you with two files, `LocalEntryPoint` (when the project runs locally), and `LambdaEntryPoint` (when deployed as lambda).

Everything works fine when running the Web API locally (LocalEntryPoint) but throws the below exception when running in lambda environment.
``System.InvalidOperationException: Unable to resolve service for type 'System.Func`1[Sentry.IHub]' while attempting to activate 'Sentry.AspNetCore.SentryTracingMiddleware'.``

With a little research I found out it's the addition I did in `Startup.cs` `Configure` method with `app.UseSentryTracing();` per documentation (https://docs.sentry.io/platforms/dotnet/guides/aspnetcore/).
`SentryTracingMiddleware` will request `IHub` with a `Func<IHub>` ctor argument which seems like a factory and not sure where to dig into finding more about it (I guess you guys add this dependency in some `IServiceCollection` extension.

Running the project locally I can confirm that `IHub` is registered in `IServiceCollection`.

How I enable Sentry.

```
public static class WebHostBuilderExtensions
    {
        public static IWebHostBuilder UseSentryIfNotEightDevelopment(
            this IWebHostBuilder webHostBuilder,
            Action<SentryAspNetCoreOptions> sentryOptionsAction = null)
        {
            var isEnabled = SentrySdk.IsEnabled;
            if (!isEnabled)
            {
                var environment = Environment.GetEnvironmentVariable("ASPNETCORE_ENVIRONMENT");
                var isDevelopment = environment is 
                    EightEnvironment.Dev or 
                    EightEnvironment.Test or 
                    EightEnvironment.QA;

                if (!isDevelopment)
                {
                    webHostBuilder.UseSentry(options => ConfigureSentryOptions(options, sentryOptionsAction));
                }
            }

            return webHostBuilder;
        }
        
        private static void ConfigureSentryOptions(SentryAspNetCoreOptions options, Action<SentryAspNetCoreOptions> sentryOptionsAction = null)
        {
            var isEnabled = SentrySdk.IsEnabled;
            if (!isEnabled)
            {
                if (sentryOptionsAction.IsNotNull())
                {
                    sentryOptionsAction(options);
                }
            }
        }
    }

```

And using the extension in LambdaEntryPoint.

```
public class LambdaEntryPoint :
        // When using an ELB's Application Load Balancer as the event source change 
        // the base class to Amazon.Lambda.AspNetCoreServer.ApplicationLoadBalancerFunction
        Amazon.Lambda.AspNetCoreServer.APIGatewayProxyFunction
    {
        /// <summary>
        /// The builder has configuration, logging and Amazon API Gateway already configured. The startup class
        /// needs to be configured in this method using the UseStartup<>() method.
        /// </summary>
        /// <param name="builder"></param>
        protected override void Init(IWebHostBuilder builder)
        {
            builder
                .UseDefaultServiceProvider(options => options.ValidateScopes = false)
                .UseSentryIfNotEightDevelopment(ConfigureSentryOptions) 
                .UseStartup<Startup>();
        }

        private static void ConfigureSentryOptions(SentryAspNetCoreOptions options)
        {
            options.Release = typeof(LambdaEntryPoint).GetTypeInfo().Assembly.GetName().Version?.ToString();
        }
    }
```

```
public void Configure(
            IApplicationBuilder app,
            IUsersAppSettings appSettings,
            IConfiguration configuration)
        {
            app.UseForwardedHeaders(new ForwardedHeadersOptions
            {
                ForwardedHeaders = ForwardedHeaders.XForwardedFor |
                                   ForwardedHeaders.XForwardedProto
            });

            if (!HostingEnvironment.IsEightDevelopment())
            {
                app.UseHsts();
                app.UseHttpsRedirection();
            }

            app.UseExceptionHandlerMiddleware();
            app.UseCancellationTokenRequestLinkTimeoutMiddleware();

            app.UseSwaggerDocuments();

            app.UseRouting();
            if (SentrySdk.IsEnabled)
            {
                app.UseSentryTracing(); // Change
            }
            app.UseAuthentication();
            app.UseAuthorization();
            app.UseCors();
            app.UseEndpoints(endpoints => { endpoints.MapControllers(); });
        }
```

**Comment by mattjohnsonpint** [comment]
Hi.  The service it's looking for is registered with everything else when you call `UseSentry`.  I think it may be getting caught up on the checks with `SentrySdk.IsEnabled`, which shouldn't be used at the registration stage.  Can you please try removing those and see if the issue persists?  Thanks.

Also, do you have anything for sentry configured in `appsettings.json`?

**Comment by weare8** [comment]
Hi, this is what I have in appsettings.json.

```
"Sentry": {
    "Dsn": "https://YYY@XXX.ingest.sentry.io/IDDD",
    "IncludeActivityData": true,
    "IncludeRequestPayload": true,
    "MinimumBreadcrumbLevel": "Error",
    "MinimumEventLevel": "Error",
    "FlushOnCompletedRequest": true,
    "MaxRequestBodySize": "Always",
    "SendDefaultPii": false,
    "AttachStackTrace": true,
    "Debug": true,
    "DiagnosticsLevel": "Error"
  }
```

I tried without the check if SDK is enabled and failed with same error.

**Comment by mattjohnsonpint** [comment]
I attempted to reproduce the problem and everything appeared to work correctly without error, using either the exact code you provided, or a simplified form.

The only way I was able to get it to error is if the call to `webHostBuilder.UseSentry` was omitted.  In that case, the call to `app.UseSentryTracing` will indeed give an error response, with the exception message you mentioned showing in the AWS CloudWatch logs.

You might take another look at your `isDevelopment` check, since you're using that to decide whether to call `UseSentry`.  Perhaps the environment variable is not what you think it is.

Either way, I'd remove the `SentrySdk.IsEnabled` test in all locations.  Instead, use the same environment check for `UseSentryTracing` as you do for `UseSentry`.
